File: python/processor.py
from pixel import Pixel
from segment import Segment
import numpy as np


def get_neighbors(point):
    """Return neighboring points"""
    r, c = point
    deltas = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (1, 1), (-1, 1), (1, -1)]
    neighbors = [(r + delta_r, c + delta_c) for delta_r, delta_c in deltas]
    return neighbors


def extract_segment(semseg, point, segment_id):
    """Uses tree search (dfs) to explore points within the segment.
    Adds all points within the segment as members of the segment and stores points at the border
    of the segment as contour points.

    Parameters:
    -----------
    semseg: ndarray
        semantic segmentation of shape (H x W x NumClasses)
    point: tuple(x, y)
        start point of dfs
    id: float
        segment id

    returns: Segment
        image segment
    """

    def within_bounds(p):
        """Checks whether point is within frame"""
        r, c = p
        h, w = semseg.shape[:2]
        return all([r >= 0, r < h, c >= 0, c < w])

    # Get label of segment
    label = semseg[point].argmax()
    # Define new segment object
    segment = Segment(label, segment_id)
    # Stack for dfs
    stack = [point]
    while stack:
        point = stack.pop()
        label, score = semseg[point].argmax(), semseg[point].max()
        # Add point as member of segment
        segment.add_pixel(point, label, score)
        for neighbor in get_neighbors(point):
            # If neighbor is out of frame or of other class than segment => contour point
            if not within_bounds(neighbor) or semseg[neighbor].argmax() != segment.label:
                segment.add_contour_pixel(point)
            # Add to stack if neighbor has not been visited
            elif not segment.seen(neighbor):
                stack.append(neighbor)

    return segment


def extract_segments(semseg):
    """Extracts image segments from semantic segmentation.

    Parameters:
    -----------
    semseg: ndarray
        semantic segmentation of shape (H x W x NumClasses)

    returns: list(Segment)
        list of image segments
    """
    segments = []
    print("Segment\t\tLabel\t\tPixels\t\tScore")
    for r in range(semseg.shape[0]):
        for c in range(semseg.shape[1]):
            # If the pixel does not belong to a known segment => extract new segment
            if not any([p.seen((r, c)) for p in segments]):
                s = extract_segment(semseg, (r, c), len(segments))
                print("%d\t\t%d\t\t%d\t\t%f" % (len(segments), s.label, len(s.pixels), s.score))
                segments.append(s)

    return segments

File: c++/src/processor.cpp
#include "processor.hpp"
#include <stack>

vector<vector<int>> getNeighbors(int x, int y) {
  vector<vector<int>> deltas = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}, {-1, -1}, {1, 1}, {-1, 1}, {1, -1}};
  vector<vector<int>> neighbors(deltas.size(), vector<int>(2, 0));

  for(int i = 0; i < deltas.size(); i++) {
    neighbors[i][0] = x + deltas[i][0];
    neighbors[i][1] = y + deltas[i][1];
  }
  return neighbors;
}

Segment* extractSegment(Semseg& semseg, int x, int y, int segmentId) {
  // Extract label and score at coordinate
  MaxActivation act = semseg.getLabelAndScore(x, y);
  int label = act.label;
  Segment* segment = new Segment(label, segmentId, semseg.getRows(), semseg.getCols());

  // Intialize stack with point (x, y)
  stack<vector<int>> s;
  s.push(vector<int>{x, y});

  // Expand for segment and save contour points
  while(!s.empty()) {
    vector<int> point = s.top();
    s.pop();
    int x = point[0], y = point[1];
    MaxActivation act = semseg.getLabelAndScore(x, y);
    segment->addPixel(x, y, act.label, act.score);

    for(vector<int> neighbor : getNeighbors(x, y)) {
      int nx = neighbor[0], ny = neighbor[1];

      // Add as contour point if new point is out of bounds
      if(!semseg.boundaryCheck(nx, ny)) {
        segment->addContourPixel(x, y, act.label, act.score);
        continue;
      }
      // Extract label and score at neighboring point
      MaxActivation nAct = semseg.getLabelAndScore(nx, ny);

      // Label missmatch => contour point
      if(nAct.label != segment->getLabel())
        segment->addContourPixel(x, y, act.label, act.score);
      else if(!segment->seen(nx, ny)) {
        s.push(neighbor);
      }
    }
  }
  return segment;
}

vector<Segment*> extractSegments(Semseg& semseg) {
  vector<Segment*> segments;
  printf("Segment\t\tLabel\t\tPixels\t\tScore\n");
  for(int y = 0; y < semseg.getRows(); y++) {
    for(int x = 0; x < semseg.getCols(); x++) {
      if(unexploredPoint(segments, x, y)) {
        int segmentId = segments.size();
        Segment* s = extractSegment(semseg, x, y, segmentId);
        segments.push_back(s);
        printf("%d\t\t%d\t\t%d\t\t%f\n", segmentId, s->getLabel(), (int)s->getSize(), s->getScore());
      }
    }
  }

  return segments;
}
